drivers: lora: Enhance CAD support

chihosin opened the pull request:
Added channel activity detection (CAD) functionality to the LoRa driver, including a maximum random delay before detection. Introduced a new structure for transmission parameters and improved memory management for the transmission buffer.

[Semtech AN1200.85](https://www.semtech.com/uploads/technology/LoRa/cad-ensuring-lora-packets.pdf)

Signed-off-by: ChihoSin chihosin@icloud.com

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -249,9 +294,36 @@ int sx12xx_lora_send_async(const struct device *dev, uint8_t *data,
-	Radio.SetMaxPayloadLength(MODEM_LORA, data_len);
+	if (dev_data.tx_cfg.cad) {
+
+		uint8_t *buffer = k_malloc(data_len);
```
No `k_malloc/k_free` in the driver please, since that adds a dependency on the system heap that does not currently exist.
You are already happy to add `k_sleep(K_MSEC(delay));` in the `lora_send_async` function, why is waiting for CAD to finish a deal breaker?

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -249,9 +294,36 @@ int sx12xx_lora_send_async(const struct device *dev, uint8_t *data,
+
+		Radio.Standby();
+
+		uint32_t delay = 20;
```
Why is there a hard limit of 20ms? (and also not documented in the header)

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -87,6 +87,12 @@ struct lora_modem_config {
+	bool cad;
+
+	/** Maximum random delay in milliseconds before channel activity detection */
+	uint32_t cad_delay_max;
```
Why is the random delay part of the API at all? Surely if someone wants this functionality that can simply delay before calling the TX function?

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -153,6 +174,24 @@ static void sx12xx_ev_rx_done(uint8_t *payload, uint16_t size, int16_t rssi,
+
+		if (modem_release(&dev_data)) {
+			/* Raise signal if provided */
+			if (sig) {
```
Silently not sending data if CAD fails without a signal being provided does not seem like a great idea.
I would propose enforcing back at the `lora_send_async` function that a signal is provided if CAD is enabled

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -9,13 +9,20 @@
+#if DT_HAS_COMPAT_STATUS_OKAY(semtech_sx1261) || \
+	DT_HAS_COMPAT_STATUS_OKAY(semtech_sx1262) || \
+	DT_HAS_COMPAT_STATUS_OKAY(st_stm32wl_subghz_radio)
+#include <sx126x-board.h>
```
I would define a helper here to simplify the `#if` further down.
```
#define SX126X_RADIO 1
```

JordanYates commented on drivers/lora/sx12xx_common.c:
```diff
@@ -228,6 +267,12 @@ int sx12xx_lora_send(const struct device *dev, uint8_t *data,
 	 * modem and driver.
 	 */
 	ret = k_poll(&evt, 1, K_MSEC(2 * air_time));
+	if (done.result < 0) {
```
`done.result` is not guaranteed to be in any particular state if `k_poll` times out.
You should be using `k_poll_signal_check`

chihosin committed:
drivers: lora: Enhance CAD support

Added channel activity detection (CAD) functionality to the LoRa driver, including a maximum random delay before detection. Introduced a new structure for transmission parameters and improved memory management for the transmission buffer.

Signed-off-by: ChihoSin chihosin@icloud.com

chihosin committed:
drivers: lora: Add CAD time calculation for transmission

This change introduces a new function to calculate the time required for a CAD operation based on the number of symbols, bandwidth, and data rate. The calculated CAD time is now incorporated into the transmission polling logic to improve timing accuracy.

Signed-off-by: ChihoSin chihosin@icloud.com

JordanYates reviewed (commented):
The more I think about this the less I like jamming it into the send functions.
CAD is a distinct operation that can also be performed without any intention to send.
My preference is to expose this as a dedicated function which returns the result of the CAD check. The user can then decide whether to proceed with the transmission, drop the transmission, or try again later.
@martinjaeger thoughts?

chihosin commented:
> The more I think about this the less I like jamming it into the send functions. CAD is a distinct operation that can also be performed without any intention to send. My preference is to expose this as a dedicated function which returns the result of the CAD check. The user can then decide whether to proceed with the transmission, drop the transmission, or try again later. @martinjaeger thoughts?

I have done this before by introducing a dedicated CAD function, but this approach introduces greater delays under high-frequency transmission and reception.

#76916

chihosin commented:
> The more I think about this the less I like jamming it into the send functions. CAD is a distinct operation that can also be performed without any intention to send. My preference is to expose this as a dedicated function which returns the result of the CAD check. The user can then decide whether to proceed with the transmission, drop the transmission, or try again later. @martinjaeger thoughts?

@JordanYates Like this #76916 ? 

I have implemented it this way before, but this approach performs unnecessary operations during the LBT process, such as the modem repeatedly acquiring and checking states or unnecessarily entering sleep mode, which introduces additional delays.

Or combining both is also a good option.

The porting work for SWL2001 is already being handled by someone else, and I’m unsure whether it’s worth continuing to improve the LoRa driver based on LoRaMac-node. The current functionality of this driver is too limited compared to RadioLib.
